File: swi_prolog/map_coloring.pl
/*

  Map coloring in SWI Prolog

  Simple map coloring problem of belgium, denmark, france, germany, netherlands, and
  luxembourg.
  go2/0 is an optimization problem.
  
  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my SWI Prolog page: http://www.hakank.org/swi_prolog/

*/

:- use_module(library(clpfd)).
:- use_module(hakank_utils).

%%
%% Plain coloring
%%
go :-
        connections(Connections), 
        NumColors = 4,
        map_color(Connections, Countries, NumColors),
        writeln(coloring=Countries),

        findall(Countries2, map_color(Connections, Countries2, NumColors),All),
        length(All,Len),
        format("It was ~d different solutions using ~d colors:\n", [Len,NumColors]),
        writeln(All).


%%
%% optimize the number of used colors
%%
go2 :-
        connections(Connections), 
        NumColors = 15,
        map_color2(Connections, Countries, NumColors, MinColors),
        writeln(Countries),
        format("We used ~d colors\n", [MinColors]).

%%
%% Simple coloring.
%%
map_color(Connections, Countries, NumColors) :-

        length(Connections,N),
        length(Countries,N),
        Countries ins 1..NumColors,
        
        findall([A,B], (between(1,N,A),
                        between(1,A,B),
                        nth1(A,Connections,Conn),
                        nth1(B,Conn,1)
                       ),
                ABs),
        maplist(not_same_color(Countries),ABs),
        
        %% symmetry breaking
        element(1,Countries,1),
        element(2,Countries,C2),
        C2 #=< 2,

        label(Countries).

not_same_color(Countries,[A,B]) :-
        element(A,Countries,CA),
        element(B,Countries,CB),
        CA #\= CB.

%%
%% Optimization: minimize the number of colors needed.
%%
map_color2(Connections, Countries, NumColors, MinColors) :-

        length(Connections,N),
        
        length(Countries,N),
        Countries ins 1..NumColors,
        MinColors in 1..NumColors, %% to optimize

        %% minimize the max number of color
        max_list_clp(Countries,MinColors),
        
        findall([A,B], (between(1,N,A),
                        between(1,A,B),
                        nth1(A,Connections,Conn),
                        nth1(B,Conn,1)
                       ),
                ABs),
        maplist(not_same_color(Countries),ABs),
        
        %% symmetry breaking
        element(1,Countries,1),
        element(2,Countries,C2),
        C2 #=< 2,

        labeling([min(MinColors)], Countries).


%%
%% Connections between these countries:
%% [belgium, denmark, france, germany, netherlands, luxembourg]
connections(A) :- 
        A = [[0, 0, 1, 1, 1, 1],
             [0, 0, 0, 1, 0, 0],
             [1, 0, 0, 1, 1, 0],
             [1, 1, 1, 0, 1, 1],
             [1, 0, 1, 1, 0, 0],
             [1, 0, 0, 1, 0, 0]].
     

File: julia/constraints/map_coloring.jl
#=

  Map coloring in Julia ConstraintSolver.jl 

  Simple map coloring problem.


  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/

=# 

using ConstraintSolver, JuMP
using Cbc, GLPK, Ipopt
const CS = ConstraintSolver
include("constraints_utils.jl")

function map_coloring(problem,print_solutions=true,all_solutions=true)

    cbc_optimizer = optimizer_with_attributes(Cbc.Optimizer, "logLevel" => 0)
    glpk_optimizer = optimizer_with_attributes(GLPK.Optimizer)
    ipopt_optimizer = optimizer_with_attributes(Ipopt.Optimizer)

    model = Model(optimizer_with_attributes(CS.Optimizer,   "all_solutions"=> all_solutions,
                                                            # "all_optimal_solutions"=>all_solutions, 
                                                            "logging"=>[],

                                                            "traverse_strategy"=>:BFS,
                                                            # "traverse_strategy"=>:DFS,
                                                            # "traverse_strategy"=>:DBFS,

                                                            # "branch_split"=>:Smallest,
                                                            # "branch_split"=>:Biggest,
                                                            "branch_split"=>:InHalf,

                                                            # https://wikunia.github.io/ConstraintSolver.jl/stable/options/#branch_strategy-(:Auto)
                                                            "branch_strategy" => :IMPS, # default
                                                            # "branch_strategy" => :ABS, # Activity Based Search
                                                            # "activity.decay" => 0.999, # default 0.999
                                                            # "activity.max_probes" => 10, # default, 10
                                                            # "activity.max_confidence_deviation" => 20, # default 20

                                                            # "simplify"=>false,
                                                            # "simplify"=>true, # default

                                                            "time_limit"=>6,

                                                            # "backtrack" => false, # default true
                                                            # "backtrack_sorting" => false, # default true

                                                            # "lp_optimizer" => cbc_optimizer,
                                                            "lp_optimizer" => glpk_optimizer,
                                                            # "lp_optimizer" => ipopt_optimizer,
                                        ))

    n, _ = size(problem)
    num_colors = 4
    # println("n:$n num_colors:$num_colors")
    @variable(model, 1 <= x[1:n] <= num_colors, Int)

    for c1 in 1:n, c2 in 1:n 
        if problem[c1,c2] == 1 
            @constraint(model,x[c1] != x[c2])
        end 
    end

    # symmertry breaking
    for c in 1:min(n,num_colors)
        @constraint(model,x[c] <= c)
    end

    # @objective(model,Max,z)

    # Solve the problem
    optimize!(model)

    status = JuMP.termination_status(model)
    # println("status:$status")
    if status == MOI.OPTIMAL
        num_sols = MOI.get(model, MOI.ResultCount())
        println("num_sols:$num_sols\n")
        if print_solutions
            for sol in 1:num_sols
                # println("solution #$sol")
                x_val = convert.(Integer,JuMP.value.(x; result=sol))
                println("x:$x_val")

            end
        end
    else
        println("status:$status")
    end

    return status
end

map_coloring_problems = Dict(
    # Connections between these countries:
    # [belgium, denmark, france, germany, netherlands, luxembourg]
    :1 => resize_matrix([[0, 0, 1, 1, 1, 1],
                         [0, 0, 0, 1, 0, 0],
                         [1, 0, 0, 1, 1, 0],
                         [1, 1, 1, 0, 1, 1],
                         [1, 0, 1, 1, 0, 0],
                         [1, 0, 0, 1, 0, 0]])
       
)

@time map_coloring(map_coloring_problems[:1])

# random instance
function random_map(n=10,lim=0.5)
    m = zeros(Int8,n,n)
    for i in 1:n, j in 1:i-1
       if rand() >= lim
        m[i,j] = 1 
        m[j,i] = 1
       end
    end
    return m
end

for n in 10:20
    println("\nn:$n")
    m = random_map(n,0.5)
    # display(m)
    @time map_coloring(m,true,false)
end
